### PR TITLE
Traverse deeply nested lists when searching for null violations

### PR DESCRIPTION
**Pull request by mattbaker**
Attempts to fix https://github.com/absinthe-graphql/absinthe/issues/1274 by traversing through layers of non-null/list wrappers when looking for non-null constraint violations.

### CODE REVIEW

File: test/absinthe/phase/execution/non_null_test.exs [hunk 1]
```diff
@@ -217,6 +240,51 @@ defmodule Absinthe.Phase.Document.Execution.NonNullTest do
...
+      {
+        deeplyNestedNonNullableListOfNullable
+      }
+      """
+
+      assert {:ok, %{data: %{"deeplyNestedNonNullableListOfNullable" => [[[["ok", nil]]]]}}} ==
+               Absinthe.run(doc, Schema)
+    end
+
+    test "deeply nested non-nullable value inside non-nullable lists cannot be null" do
```
Comment (mattbaker) [inline test/absinthe/phase/execution/non_null_test.exs:265]: This wasn't the issue I saw when I started digging into this, but this test also fails on master. I think it's in the same theme of needing to traverse through multiple layers of lists/non-nulls.

File: test/absinthe/phase/execution/non_null_test.exs [hunk 2]
```diff
@@ -217,6 +240,51 @@ defmodule Absinthe.Phase.Document.Execution.NonNullTest do
   end
 
   describe "lists" do
+    test "non-null list of non-null list of nullable value returns null value" do
```
Comment (mattbaker) [inline test/absinthe/phase/execution/non_null_test.exs:243]: This seems like it should be redundant when `test "deeply nested nullable value inside non-nullable lists can be null"` exists below, but the latter actually passes in master for the same reason `test "deeply nested non-nullable value inside non-nullable lists cannot be null"` fails (I think). 

So I ended up keeping this test just for my own sanity.

**Comment by mattbaker** [comment]
@benwilson512: Full transparency — this is a complex file and I don't know if what I'm doing is the right thing here, it could definitely benefit from your eyes!

It seems like, one way or another, this boils down to the fact that a sufficiently deep wrapping of `non_null(list_of(non_null(list_of(......` will cause issues? I took a stab at making the "add errors" codes and the null propagation code more recursive and it seemed to do the trick. My only worry is I feel like I don't fully grok everything in this module and I may have done the wrong thing and just gotten lucky 😱

**Comment by mattbaker** [comment]
@benwilson512 - Curious if you have any thoughts, but I'm also sure you're busy and I know this is probably more than a surface fix! I could try to set some time aside this month to dive deeper and build some more familiarity myself and maybe emerge with a bit more confidence too, if you think you won't have time to take a look. It's also not super urgent for us.

File: lib/absinthe/phase/document/execution/resolution.ex [hunk 3]
```diff
@@ -285,30 +285,28 @@ defmodule Absinthe.Phase.Document.Execution.Resolution do
...
-       when is_list(values) do
-    values
-    |> Enum.with_index()
-    |> Enum.filter(&is_nil(elem(&1, 0)))
-    |> Enum.map(fn {_value, index} ->
-      %{message: "Cannot return null for non-nullable field", path: [index]}
+  defp maybe_add_non_null_error([], [_ | _] = values, %Type.List{of_type: type}, path) do
+    Enum.with_index(values)
+    |> Enum.flat_map(fn {value, index} ->
+      maybe_add_non_null_error([], value, type, path ++ [index])
```
Comment (benwilson512) [inline lib/absinthe/phase/document/execution/resolution.ex:305]: Appending to this list here will be `O(n^2)` and the lists could potentially be quite large. I'm working to refactor this.

**Review by benwilson512 — APPROVED** [review]
Sorry this took me so long to get to, thanks Matt!

File: lib/absinthe/phase/document/execution/resolution.ex (hunk 3 above)
Comment (mattbaker) [reply lib/absinthe/phase/document/execution/resolution.ex]: Oh very good catch, I breezed right past that!

**Comment by mattbaker** [comment]
Not a problem at all! It was an odd edge case that was on our "to do" list but wasn't urgent at all. 

Thanks so much!

**Pull request merged**